### PR TITLE
HTTP/3 connection params

### PR DESCRIPTION
**Pull request by eagr**
Dedup client/server builders. And it's likely there will be more shared parameters in the future.

### CODE REVIEW

**Comment by eagr** [comment]
Should we keep the client/server builders? Also, any suggestions on the name of the parameter builder?

**Comment by Ruben2424** [comment]
> Should we keep the client/server builders?

I would keep the client/server builders because there will possibly be parameters which only belong to server or client.

>  Also, any suggestions on the name of the parameter builder?

Maybe just `ParameterBuilder`? But i'm not good with names.

File: h3/src/params.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        }
+    }
+}
+
+impl Params {
+    /// Default max header size
+    pub const DEFAULT_MAX_FIELD_SECTION_SIZE: u64 = (1 << 62) - 1;
+
+    /// Set whether WebTransport is supported
+    pub fn enable_webtransport(mut self, val: bool) -> Self {
```
Comment (Ruben2424) [inline h3/src/params.rs:25]: It might be better to put the WebTransport stuff behind a Feature Flag. What do you think?

Comment (eagr) [reply h3/src/params.rs:25]: Yeah, I thought about that too. But what if it's just for params and settings?

Comment (Ruben2424) [reply h3/src/params.rs:25]: I would put WebTransport completely behind a feature flag or maybe even in a separate crate. Iirc there was some discussion about this in #71.

Comment (eagr) [reply h3/src/params.rs:25]: I understand the benefits of putting stuff behind a feature gate sometimes (smaller binary, faster compilation, etc). I totally agree we should feature-gate or separate major implementations of WT.

But configs and settings are lightweight and hard to separate. The benefit of putting those behind a feature gate seems limited. I'm 100% sure about this though. What do you think? @seanmonstar

Comment (seanmonstar) [reply h3/src/params.rs]: I have not had the time to read the WebTransport spec, so not sure what more is needed, but: does adding this at least allow for WebTransport to be _proxied_? If it's needed for a proxy to simply pass the data frames along, then seems fair to include it as part of the crate, perhaps?

Comment (eagr) [reply h3/src/params.rs]: Yeah, I think so. The setting would be required to be sent by both sides when establishing connection to create a WT session.

**Comment by eagr** [comment]
The API still feels a bit weird. There should probably be one builder instead of two: `Config` and `Builder`.

```rs
h3::config::Config {}

h3::client::Builder { config }
h3::server::Builder { config }
```

So, it should either be something that looks like

```rs
// somehow dedupped
h3::config::ClientConfig {}
h3::config::ServerConfig {}

// no more Builder
h3::client::new(conn, client_config)
h3::server::new(conn, server_config)
```

or

```rs
// somehow dedupped
h3::Endpoint::client().build(conn)
h3::Endpoint::server().build(conn)
```

**Comment by Ruben2424** [comment]
I like this option better.
```rs
// somehow dedupped
h3::config::ClientConfig {}
h3::config::ServerConfig {}

// no more Builder
h3::client::new(conn, client_config)
h3::server::new(conn, server_config)
```

**Comment by eagr** [comment]
Yeah. Using a config struct has a potential advantage of easier mapping to settings. We could potentially have something like `impl From<Config> for Settings`.

**Comment by eagr** [comment]
Changed user-facing config structs to be compositions: 

```rs
ClientConfig { conn: ConnectionConfig, ... }
ServerConfig { conn: ConnectionConfig, ... }
```

this way we could could easily provide some config for the shared `ConnectionInner`.

**Pull request closed without merging**